### ISSUE TITLE
FormData in POST is treated as JSON

### ISSUE THREAD

**Issue by bep** [marmelab/restful.js#72]
It should be `multipart/form-data ` with content-type and boundaries set by the browser.

Doing a `fetch` POST with the same form object works:

```
fetch('/api', {
            method: 'post',
            body: multiformData
        });
```

See #46

**Comment by RobinBressan** [comment]
Hi,

Have you tried to set the correct `Content-Type` header? If you don't, restul.js will assume it is JSON and will set it to `application/json;charset=UTF-8` (see https://github.com/marmelab/restful.js/blob/master/src/model/endpoint.js#L25).

**Comment by bep** [comment]
Yes, if I set the correct content header the server complains about missing boundaries. I have spent some hours reading and debugging this, and in this case the trick is to let the browser handle everything (setting content header and correct boundaries between the parts).

**Comment by RobinBressan** [comment]
You can use a `requestInterceptor` to deal with the config object forwarded to fetch. This way you should be able to remove the `Content-Type` header.

**Comment by bep** [comment]
The requestInterceptor seems to be tiggered too late in the game.

```
res.addRequestInterceptor((config) => {
            const {
                data, headers, method, params, url
            } = config;
            
            delete headers["Content-Type"];
            return {
                headers
            };
        });
```

The correct content type doesn't get set by the browser.

**Comment by cristianfierro** [comment]
Any solution for this?

**Comment by sohibul** [comment]
waiting the solution too

**Comment by PM5544** [comment]
I've had a similar problem but I do not experience the same issues @bep has.
In my case the following was the case.
When sending in a standard JS object as body (like I think everyone is doing by default) it gets JSON.stringify()'d by the browser when constructing the request and it looks like restfull.js is treating it as JSON, but in fact the browser does this.
I've fixed this issue by formatting the data from the form like the browser expects it (like GET variables only in the body of the request) with a function in the requestInterceptor like so:

`res.addRequestInterceptor((params) => {
  const {data, headers} = params;
  let str = '';
  if (data){
    Object.keys(data).forEach((key) => {
      str += ( str.length ? '&' : '' ) + key + '=' + data[key];
    });
    headers['Content-Type'] = 'multipart/form-data';
  }

  return {
    headers,
    data: str
  };
});`
And experienced no other issues thus far.
Hope it helps in your cases.

PM5544

**Comment by sohibul** [comment]
@PM5544 so, you pass the data as json object?

**Comment by PM5544** [comment]
@sohibul To restfull.js yes and then it gets reformatted by the interceptor to the url-parameter format the backend expects when receiving multipart/form-data.

**Comment by bep** [comment]
@PM5544 the main point of this issue is to make sure it is handled natively and effectively as a multiform formdata stream by the browser.

It should be a simple conditional fix to this library.

**Comment by PM5544** [comment]
@bep yes i understand your point. the reason of my reply was to give others with similar issues a way to work around it until a long term solution was formed, if @RobinBressan sees the need to.

**Comment by sohibul** [comment]
@PM5544 do you have tried to send file object? I need `multipart/form-data` for upload file

**Comment by PM5544** [comment]
@sohibul nope I have not, but I assume it's better left up to the browser to directly connect to an upload service on the backend by using a standard form with an input with type="file".
